### PR TITLE
Update development/v0.1.4 branch with diagram creation using textual representation

### PR DESCRIPTION
**Pull request by chuajiaxuan**

### CODE REVIEW

**Comment by amoshydra** [comment]
I will do a manual merging with `dev/encapsulate-viewer-with-iframe` and submit a pull-request together.
Putting content into iframe will change how some of the library behave.

**Comment by amoshydra** [comment]
This pull-request is closed. Do not delete the branch yet.

File: sashimi-webapp/src/logic/documentPackager/markdown-it-diagram.js [hunk 1]
```diff
@@ -6,12 +6,32 @@
...
+}
+
 export default function setup(md, options) {
   const defaultRender = md.renderer.rules.fence;
 
   /* eslint no-undef: 0 */
   md.renderer.rules.fence = (tokens, idx, opts, env, self) => {
     const token = tokens[idx];
+    token.content = filter(token.content);
     if (token.info === 'sequence') {
```
Comment (amoshydra) [inline sashimi-webapp/src/logic/documentPackager/markdown-it-diagram.js]: Looks like switch case could be beneficial here to remove the need of repeating `<pre class="sequence">${token.content}</pre>`;


You can consider doing this:

```javascript
switch (token.info) {
  case 'sequence':
  case 'flow':
  case 'mermaid':
  case 'graphviz':  // fallthrought
    return `<pre class="${token.info}">${token.content}</pre>`;
  default: {
    // pass token to default renderer.
    return defaultRender(tokens, idx, opts, env, self);
  }
}
```

**Comment by amoshydra** [comment]
Manual merging is not necessary, using iframe for rendering does not change any behaviour. 
We can merge this after reviewing. Reopening this pull-request

File: sashimi-webapp/src/logic/renderer/diagrams.js [hunk 2]
```diff
@@ -21,31 +21,32 @@ export default function diagramsRenderer(ele) {
...
       });
-
       let content = seqDiagrams[i].innerHTML;
-      /* eslint no-useless-escape: 0*/
-      content = content.replace(/\&gt\;/g, '>');
-      seqDiagrams[i].innerHTML = '';
-      /* eslint no-undef: 0 */
+      content = content.replace(/&gt;/g, '>');
       const diagram = Diagram.parse(content);
+      seqDiagrams[i].innerHTML = '';
```
Comment (amoshydra) [inline sashimi-webapp/src/logic/renderer/diagrams.js]: Using `innerHTML` may not be a good practice here.
In the future, you may consider manipulating the DOM using api like "appendChild", "removeChild" etc.

See:
- http://stackoverflow.com/questions/3955229/remove-all-child-elements-of-a-dom-node-in-javascript
-  http://stackoverflow.com/questions/7476638/if-people-recommend-i-shouldnt-use-innerhtml-what-then-should-i-instead-use

In this case, you may consider using a for loop to remove all the children inside seqDiagrams[i]

**Review by amoshydra — APPROVED** [review]
Looks good. 
Added 2 suggestions for this pull-request. You may review the suggestion and merge the pr.

File: sashimi-webapp/src/logic/documentPackager/markdown-it-diagram.js (hunk 1 above)
Comment (chuajiaxuan) [reply sashimi-webapp/src/logic/documentPackager/markdown-it-diagram.js]: Thanks! You gave me another idea! I store the accepted diagrams library in an array and do this:
```javascript
if (diagramsLib.includes(token.info)) {
      return `<pre class="${token.info}">${token.content}</pre>`;
}
```

File: sashimi-webapp/src/logic/renderer/diagrams.js (hunk 2 above)
Comment (chuajiaxuan) [reply sashimi-webapp/src/logic/renderer/diagrams.js]: Ahh yes yes. I have read about this before! Will go change it! :)

**Pull request merged**